File: leetcode/BackTracking/leetcode-0216/216.java
class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> array = new ArrayList<>();
        helper(k, n, 1, 0, res, array);
        return res;
    }

    private void helper(int k, int n, int start, int sum, List<List<Integer>> res, List<Integer> array) {
        if (array.size() == k && sum == n) {
            res.add(new ArrayList<Integer>(array));
            return;
        }
        if (array.size() >= k) {
            return;
        }

        for (int i = start; i <= 9; i++) {
            array.add(i);
            helper(k, n, i + 1, sum + i, res, array);
            array.remove(array.size() - 1);
        }
    }
}

File: leetcode/BackTracking/leetcode-0216/216.go
package backtracking

func combinationSum3(k int, n int) [][]int {
	res := [][]int{}
	array := []int{}
	helper(k, n, 1, 0, array, &res)
	return res
}

func helper(k, n, start, sum int, array []int, res *[][]int) {
	if len(array) == k && sum == n {
		tmp := make([]int, len(array))
		copy(tmp, array)
		*res = append(*res, tmp)
		return
	}

	if len(array) >= k {
		return
	}

	for i := start; i <= 9; i++ {
		array = append(array, i)
		helper(k, n, i+1, sum+i, array, res)
		array = array[:len(array)-1]
	}
}
